ofrak_core/ofrak/core/elf/analyzer.py

```diff
@@ -239,6 +245,80 @@ def deserialize(cls, deserializer: BinaryDeserializer, elf_index: int) -> ElfSym
         return ElfSymbol(elf_index, st_name, st_value, st_size, st_info, st_other, st_shndx)
 
 
+class ElfLinkableSymbolIdentifier(Identifier[None]):
+    """
+    ElfSymbols are linkable, so tag them as LinkableSymbols.
+    """
+
+    targets = (ElfSymbol,)
+
+    async def identify(self, resource: Resource, config: None) -> None:
+        elf_sym = await resource.view_as(ElfSymbol)
+        if elf_sym.st_shndx == 0 or elf_sym.st_shndx >= 0xFF00:
+            return
+        if elf_sym.symbol_index == 0:
+            return
+        if elf_sym.st_name == 0:
+            return
+        if elf_sym.get_type() not in [ElfSymbolType.FUNC, ElfSymbolType.OBJECT]:
+            return
+        resource.add_tag(LinkableSymbol)
+
+
+class ElfLinkableSymbolAnalyzer(Analyzer[None, LinkableSymbol]):
+    """
+    Extract the linking info relevant for OFRAK's PatchMaker.
+    """
+
+    targets = (ElfSymbol,)
+    outputs = (LinkableSymbol,)
+
+    async def analyze(self, resource: Resource, config: None) -> LinkableSymbol:
+        elf_symbol = await resource.view_as(ElfSymbol)
+        elf_r = await resource.get_only_ancestor(ResourceFilter(tags=(Elf,)))
+        prog_attrs = await elf_r.analyze(ProgramAttributes)
+
+        sym_name = await elf_symbol.get_name()
+        if elf_symbol.get_type() is ElfSymbolType.FUNC:
+            sym_type = LinkableSymbolType.FUNC
+        elif elf_symbol.get_type() is ElfSymbolType.OBJECT:
+            # Need to check what section it is in to see if it is writable
+            section_containing_symbol = await elf_r.get_only_descendant_as_view(
+                ElfSectionHeader,
+                r_filter=ResourceFilter(
+                    tags=(ElfSectionHeader,),
+                    attribute_filters=(
+                        ResourceAttributeValueFilter(
+                            ElfSectionHeader.SectionIndex, elf_symbol.st_shndx
+                        ),
+                    ),
+                ),
+            )
```

Comment:
I had the same reaction -- it seems like it would be much more efficient to extract this information inside of the `ElfSymbolUnpacker` (getting the `sym_type` would only need to be done once per section). The same is true for `ComplexBlock` -- the linkable symbol information can be extracted during unpacking.

This lead to me thinking about the relationship/repetition between `LinkableSymbol`, `ComplexBlock.Symbol`, and `ElfSymbol`. With the current implementation, it is highly likely that there can be two `LinkableSymbol` views for the same virtual address (one from disassembler, one from Elf) -- there is, however, no information about where the `LinkableSymbol` came from.

It seems like, at a minimum, it would be helpful to know:
- Did this symbol come from a disassembler backend?
- Did this symbol come from Elf information?
- Was this symbol manually defined?

I wonder if `LinkableSymbol` is really an abstract class that `ComplexBlock`, `ElfSymbol` (and maybe `UserAddedSymbol`?) implement.

Reply:
> I had the same reaction -- it seems like it would be much more efficient to extract this information inside of the ElfSymbolUnpacker (getting the sym_type would only need to be done once per section).

It would only be "more efficient" in the case we actually use linkable symbols, which is a small minority of the time we unpack ELFs. In the majority of cases it would be wasted work. In this implementation, we only analyze LinkableSymbols if we actually use them, so if efficiency is a concern then we need to consider what cases we are optimizing for. I don't think we should be optimizing though. If you can show me the implementation is too inefficient in a situation, then we can optimize for that.


> With the current implementation, it is highly likely that there can be two LinkableSymbol views for the same virtual address (one from disassembler, one from Elf)
It is perfectly fine and normal for a single vaddr to have multiple symbols map to it. The only real "conflict" is if one name is shared between symbols, and there is code to check for it as @andresito00 pointed out:

```
            if sym.name in symbols_by_name:
                other_sym = symbols_by_name[sym.name]
                if other_sym == sym:
                    continue
                else:
                    sym_source, _ = _describe_symbol_source(sym)
                    other_sym_source, _ = _describe_symbol_source(other_sym)

                    raise ValueError(
                        f"Multiple symbols with the name {sym.name} found, and they have different "
                        f"information! {sym} ({sym_source}) vs. {other_sym} ({other_sym_source})"
                    )
```

>It seems like, at a minimum, it would be helpful to know:
> 
> Did this symbol come from a disassembler backend?
> Did this symbol come from Elf information?
> Was this symbol manually defined?

I nearly added this as a new field of LinkableSymbol, before I remember that OFRAK always tracks that info for us! 

```
    sym_source = cast(
        bytes,
        sym.resource.get_model().get_component_id_by_attributes(LinkableSymbol.attributes_type),
    )
    if sym_source == CLIENT_COMPONENT_ID:
        sym_source_str = f"defined manually by adding view or attributes in a script"
    elif sym_source == UpdateLinkableSymbolsModifier.id:
        sym_source_str = f"defined manually via {sym_source.decode('ascii')}"
    else:
        sym_source_str = f"analyzed via {sym_source.decode('ascii')}"
```

>I wonder if LinkableSymbol is really an abstract class that ComplexBlock, ElfSymbol (and maybe UserAddedSymbol?) implement.

I don't think so. `LinkableSymbol` inherits from `LabeledAddress`, which I think could be a superclass of `ComplexBlock`, and that was the original intention. But in general I'm becoming less inclined to fit everything into one inheritance tree. 

I think that how we design tags should be more oriented around the "domain" that they are concerned with. In this case, `ComplexBlock` tag is part of the binary/disassembly structure "domain", and `ElfSymbol` is part of the ELF structure "domain", and `LinkableSymbol` is part of the code patching "domain". Yes, these are all intertwined, but that's why are always working with the same underlying resources throughout these domains. I'd rather these resources cross these domains at pretty defined points (the identifiers/analyzers in this PR) rather than all the tags be all tangled up. I think I'll even move the Identifiers and Analyzers to be with the other linkable_* code, rather than where they currently are now (grouped with ELF or complex block code, respectively).